fvm/crypto/crypto_test.go

```diff
@@ -0,0 +1,229 @@
+		correctCombinations[runtime.SignatureAlgorithmECDSA_secp256k1][runtime.HashAlgorithmSHA2_256] = struct{}{}
+		correctCombinations[runtime.SignatureAlgorithmECDSA_secp256k1][runtime.HashAlgorithmSHA3_256] = struct{}{}
+
+		signatureAlgos := []runtime.SignatureAlgorithm{
+			runtime.SignatureAlgorithmECDSA_P256,
+			runtime.SignatureAlgorithmECDSA_secp256k1,
+			runtime.SignatureAlgorithmBLS_BLS12_381,
+		}
+		hashAlgos := []runtime.HashAlgorithm{
+			runtime.HashAlgorithmSHA2_256,
+			runtime.HashAlgorithmSHA2_384,
+			runtime.HashAlgorithmSHA3_256,
+			runtime.HashAlgorithmSHA3_384,
+			runtime.HashAlgorithmKMAC128_BLS_BLS12_381,
+		}
+
+		for _, s := range signatureAlgos {
+			for _, h := range hashAlgos {
+				t.Run(fmt.Sprintf("combination: %v, %v", s, h), func(t *testing.T) {
+					seed := make([]byte, 256)
+					rand.Read(seed)
+					pk, err := gocrypto.GeneratePrivateKey(crypto.RuntimeToCryptoSigningAlgorithm(s), seed)
+					require.NoError(t, err)
+
+					tag := string(flow.UserDomainTag[:])
+					var hasher hash.Hasher
+					if h != runtime.HashAlgorithmKMAC128_BLS_BLS12_381 {
+						hasher, err = crypto.NewPrefixedHashing(crypto.RuntimeToCryptoHashingAlgorithm(h), tag)
+						require.NoError(t, err)
+					} else {
+						hasher = gocrypto.NewBLSKMAC(tag)
+					}
+
+					signature := make([]byte, 0)
+					sig, err := pk.Sign([]byte("some data"), hasher)
+					if _, shouldBeOk := correctCombinations[s][h]; shouldBeOk {
+						require.NoError(t, err)
+					}
+
+					if sig != nil {
+						signature = sig.Bytes()
+					}
+
+					ok, err := crypto.VerifySignatureFromRuntime(
+						crypto.NewDefaultSignatureVerifier(),
+						signature,
+						tag,
+						[]byte("some data"),
+						pk.PublicKey().Encode(),
+						s,
+						h,
+					)
+
+					if _, shouldBeOk := correctCombinations[s][h]; shouldBeOk {
+						require.NoError(t, err)
+						require.True(t, ok)
+					} else {
+						require.Error(t, err)
+						require.False(t, ok)
+					}
+				})
+			}
+		}
+	})
+
+	t.Run("BLS verification tag size > 32 bytes should pass", func(t *testing.T) {
+		seed := make([]byte, 256)
+		rand.Read(seed)
+		pk, err := gocrypto.GeneratePrivateKey(gocrypto.BLSBLS12381, seed)
+		require.NoError(t, err)
+
+		tag := make([]byte, 256)
+		rand.Read(tag)
+		hasher := gocrypto.NewBLSKMAC(string(tag))
+
+		signature := make([]byte, 0)
+		sig, err := pk.Sign([]byte("some data"), hasher)
+		require.NoError(t, err)
+
+		if sig != nil {
+			signature = sig.Bytes()
+		}
+
+		ok, err := crypto.VerifySignatureFromRuntime(
+			crypto.NewDefaultSignatureVerifier(),
+			signature,
+			string(tag),
+			[]byte("some data"),
+			pk.PublicKey().Encode(),
+			runtime.SignatureAlgorithmBLS_BLS12_381,
+			runtime.HashAlgorithmKMAC128_BLS_BLS12_381,
+		)
+
+		require.NoError(t, err)
+		require.True(t, ok)
+	})
+
+	t.Run("tag combinations", func(t *testing.T) {
+
+		cases := []struct {
+			hashTag   string
```

Comment:
This would be clearer.
```suggestion
			signTag   string
```